Move instead of merge symex L2 renaming map when possible

JohnDumbell opened the pull request:
This PR splits the level 2 renaming map into two: a global map whose state is kept across all symex runs, and a local one that only holds the generation counts of something along a particular branch (or specifically, a statet).

It works like this: on reading the current generation, the local map will always be used. When a generation is increased, we query the global map for the currently-taken generation number, increment that by one, then assign the local maps generation number as that. With this, we don't have to merge maps at each branch if for some reason a particular branch can't be taken, as generation numbers will always be unique. Downside is that generation numbers might look strange if symex has many paths to take.

Example of this: path symex, previously, would have taken the next free number on that particular path. With the global name map this is actually now the next free number across _all_ paths. So you could see:

name#1
name#2
name#54

Which might throw people off. You fix this by making the global name map only 'global' to an individual run of symex, but for now there shouldn't be any problems, and the change is minuscule anyway.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

smowton commented:
@tautschnig parts of this are pretty similar to https://github.com/diffblue/cbmc/pull/3980 and its separation of unique name allocation from current-name tracking -- just a heads up in case there's a danger of duplicating work, and @JohnDumbell you might want to read that PR to compare and try to align your naming and ownership of data structures with @tautschnig's choices.

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -472,6 +486,14 @@ static void merge_names(
+    !goto_state.guard.is_false(),
+    "an unreachable state should never have been queued for merging");
+
   if(dest_state.guard.is_false())
```
These should go away considering the invariants

smowton commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +86,63 @@ struct symex_level2t : public symex_renaming_levelt
+
+  /// In the case of a state split, we want to keep track of the overall
+  /// name generation across all states for identifier creation.
+  optionalt<current_namest> local_generations;
```
It looks like you're using the existing `current_names` to track global-highest-generation and this new `local_generations` to track the latest version on a particular path? If so, see #3980 -- that creates `unique_index_map` to store a global highest L1 generation, and that map belongs to `path_storaget`, not `symex_level2t` which is path-local. (as a rule: varies per path -> `goto_statet`, varies per path but doesn't need to be copied on split -> `goto_symex_statet`, doesn't vary per path -> `path_storaget`). So you should have a map `unique_l2_index_map` or something on `path_storaget` used to allocate fresh L2 names, and `current_names` should continue to describe what L2 generation is observed by a read on a particular path.

smowton commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +86,63 @@ struct symex_level2t : public symex_renaming_levelt
+  /// If we have a local counter, use that instead of our global.
+  unsigned current_count(const irep_idt &identifier) const override
+  {
+    if(local_generations)
```
This should be simpler -- the current count (i.e. the generation observed by a read) should always just be path-local-current-L2-names.at(identifier)

smowton commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +86,63 @@ struct symex_level2t : public symex_renaming_levelt
+
+  void erase(const irep_idt &identifier)
+  {
+    current_names.erase(identifier);
```
Given `current_names` is the L2 global unique generation, and the `erase` method is path-local, I think this is dubious.

JohnDumbell commented:
Sorry for the mass review auto-add. Feel free to remove yourself (as I can't), a push accidentally hoovered up loads of commits I didn't intend.

smowton commented:
I can't remove them either I'm afraid. Can kill the PR and remake, or just rely on them filtering the thread out as most reviewers do 99% of review requests ;)

tautschnig commented:
There is certainly overlap with #3980, but also #4169. While those are awaiting reviews I'd be happy to factor out any changes that may independently be useful. Otherwise we should agree on a sequence of which PRs are suppose to go in on top of which other and then just start stacking up the commits appropriately in the PRs.

JohnDumbell commented:
@tautschnig Thanks, though as the other two PR's are further along the easiest solution is probably for me to just build off those two and/or wait for them to go in first. I'll cherry-pick anything directly useful, otherwise I think my changes will sit alongside them.

tautschnig commented:
@JohnDumbell Sounds good - just let me know when this PR is ready for reviews/you'd like comments.

smowton commented on src/goto-symex/path_storage.h:
```diff
@@ -90,6 +90,12 @@ class path_storaget
   /// Counter for nondet objects, which require unique names
   symex_nondet_generatort build_symex_nondet;
 
+  /// Storage for overall level two names across an entire symex run. This will
```
I suspect `shared_ptr` is overkill, as this reference should outlive all others

JohnDumbell replied:
It is, it was just the easiest thing for copy/moving without too much hassle. Have the syntax handy to be able to pointer-ize a field properly when ripping it out of an object and putting it into a shared pointer elsewhere?

smowton commented on src/goto-symex/path_storage.h:
```diff
@@ -90,6 +90,12 @@ class path_storaget
 
+  /// Storage for overall level two names across an entire symex run. This will
+  /// be shared across all states to allocate unique generations.
+  std::shared_ptr<symex_renaming_levelt::current_namest> level2_names =
```
Might be misleading to use `current_namest`, since this is explicitly *not* a store of current names -- in fact this happening to be the same structure as current-names is just a coincidence. Use a separate typedef.

smowton commented on src/goto-symex/path_storage.h:
```diff
@@ -90,6 +90,12 @@ class path_storaget
   /// Counter for nondet objects, which require unique names
   symex_nondet_generatort build_symex_nondet;
 
+  /// Storage for overall level two names across an entire symex run. This will
```
"overall" is a bit vague -- how about "Tracks fresh L2 names across an entire symex run"

smowton commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +86,74 @@ struct symex_level2t : public symex_renaming_levelt
+  void increase_generation(const irep_idt l1_identifier, const ssa_exprt &lhs)
+  {
+    current_names.emplace(l1_identifier, std::make_pair(lhs, 0));
+    if(global_names != nullptr)
```
Surely, must never be null

JohnDumbell replied:
In all current situations - yes, but technically it's not really an absolute requirement for the operation of the object. It's an if/when scenario, but it makes it more robust at very little code and complexity increase.

Edit: I don't have a super-strong opinion either way, but I felt it added enough with such a little downside it was worth doing.

romainbrenguier replied:
add `INVARIANT(global_names != nullptr)` if it's never expected to be null.

smowton replied:
I'd say assume it's always not null and let any subsequent editor who cares about that use case add it; it'll only mould anyhow with no users.

smowton commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +86,74 @@ struct symex_level2t : public symex_renaming_levelt
+  /// if there isn't an expression keyed by the l1 identifier.
+  void increase_generation(const irep_idt identifier)
+  {
+    // If we can't find the name in the local scope, don't incr ease the global
```
typo

smowton commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +86,74 @@ struct symex_level2t : public symex_renaming_levelt
+        global_names_iter->second.second++,
+          current_names_iter->second.second = global_names_iter->second.second;
+    }
+    else
```
This should almost certainly never happen -- a fresh name should always be a global matter

smowton replied:
Per above, remove this dead code path

smowton commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +86,74 @@ struct symex_level2t : public symex_renaming_levelt
+    {
+      auto global_names_iter = global_names->find(identifier);
+      if(global_names_iter != global_names->end())
+        global_names_iter->second.second++,
```
Don't use a comma-expression (surprising)

smowton commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +86,74 @@ struct symex_level2t : public symex_renaming_levelt
+    increase_generation(l1_identifier);
+  }
+
+  /// Ages the generation of an ssa expression by one. Does nothing
```
How about "Allocates a fresh L2 name for the given L1 identifier, and makes it the latest generation on this path"

smowton commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +86,74 @@ struct symex_level2t : public symex_renaming_levelt
+  {
+    // If we can't find the name in the local scope, don't incr ease the global
+    // even if it exists there.
+    auto current_names_iter = current_names.find(identifier);
```
Is it always a bug if symex tries to allocate a new generation of something that hasn't been declared?

smowton replied:
Question still outstanding

JohnDumbell replied:
I'm not sure, but for me the question is whether highlighting that lack here is actually useful. If I did put any sort of assertion in I'd just make this private and add another public method like increase_generation_if_exists instead, so nothing would really change.

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -221,7 +221,14 @@ void goto_symext::symex_goto(statet &state)
     state.top().goto_state_map[new_state_pc];
 
-  goto_state_list.emplace_back(state);
+  if(new_guard.is_true())
```
Add a comment noting in the must-take case we're free to move the state away, since it can't be used while the guard is false

JohnDumbell committed:
Adding move constructors

JohnDumbell committed:
Add global name map to goto_symex_state

This global name map will be used to check what generation is currently
available for each level1 name, and shared across all states. This will
only be used when a particular state tries to find out what the next
free generation is.

The main draw of this is that all branches will now assign unique
generations that won't clash with assignments done across other
branches. One minor downside is that in VCC's the generation number
might jump sporadically (say from #4 to #40).

JohnDumbell committed:
Optimise symex state management for straight-line code with GOTOs

In the situation of merge on a state that hasn't been traversed (with a
false guard) try and move the names into the destination state instead
of a merge or copy operation

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: dc7a7b2772ce4e1b2c8305390c6ada5c55646730).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102225242

romainbrenguier commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -73,10 +73,29 @@ class goto_statet
+      source(std::move(other.source)),
+      propagation(std::move(other.propagation)),
+      atomic_section_id(other.atomic_section_id),
+      safe_pointers(other.safe_pointers),
```
why is this not moved?

smowton replied:
It actually shouldn't be here at all, since it doesn't vary per program point, but IIRC @tautschnig has a PR in progress to fix that.

romainbrenguier commented on src/goto-symex/path_storage.h:
```diff
@@ -90,6 +90,13 @@ class path_storaget
   /// Counter for nondet objects, which require unique names
   symex_nondet_generatort build_symex_nondet;
 
+  typedef std::map<irep_idt, std::pair<ssa_exprt, unsigned>> global_name_mapt;
```
use `using`

smowton replied:
What's the difference?

JohnDumbell replied:
Here, nothing particularly, but it's useful in very niche template inference situations where typedefs just don't work, so it's just better afaik.

romainbrenguier commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +86,77 @@ struct symex_level2t : public symex_renaming_levelt
+  /// Ages the generation of an ssa expression by one.
+  void increase_generation(const irep_idt l1_identifier, const ssa_exprt &lhs)
+  {
+    current_names.emplace(l1_identifier, std::make_pair(lhs, 0));
```
definition should be moved to the cpp file

romainbrenguier commented on src/goto-symex/renaming_level.h:
```diff
@@ -13,8 +13,10 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
+#include <memory>
 #include <unordered_set>
 
+#include <iostream>
```
not necessary if you move definitions to the cpp file

romainbrenguier commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -514,10 +523,16 @@ static void merge_names(
-void goto_symext::phi_function(
-  const goto_statet &goto_state,
-  statet &dest_state)
+void goto_symext::phi_function(goto_statet &goto_state, statet &dest_state)
```
pass by copy or && rather than reference if move is used on it in the function

smowton replied:
Agreed, recommend pass by &&

JohnDumbell replied:
It's not moved all the time though, and I'm not technically moving the entire structure either just parts of it so wouldn't that create move/copy arbitrarily on the existence of constructors with no nuance in how it's actually used?

romainbrenguier commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +86,77 @@ struct symex_level2t : public symex_renaming_levelt
+    increase_generation(l1_identifier);
+  }
+
+  /// Allocates a fresh L2 name for the given L1 identifier, and makes it the
```
:pick: Not sure "allocates" is appropriate for a name, there is no memory allocation involved.

smowton replied:
I think allocates is ok, you can allocate something other than a memory block

romainbrenguier commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +86,77 @@ struct symex_level2t : public symex_renaming_levelt
+
+  /// Prints the differences between the global and local naming maps (if they
+  /// exist)
+  void print_differences(const std::string &addition)
```
This should be marked as a debug function

smowton replied:
Probably surround with `#if 0` along with an `iostream` include (and use `std::cerr`, since it's debug output?)

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -221,7 +221,16 @@ void goto_symext::symex_goto(statet &state)
+  }
+  else
+  {
+    goto_state_list.push_back(statet::goto_statet(state));
```
Use `emplace_back` like the code you're replacing (in both cases)

smowton commented on src/goto-symex/renaming_level.h:
```diff
@@ -84,6 +85,51 @@ struct symex_level2t : public symex_renaming_levelt
+  std::shared_ptr<current_namest> global_names;
+
+  /// Ages the generation of an ssa expression by one.
+  void increase_generation(const irep_idt l1_identifier, const ssa_exprt &lhs);
```
Copy doc comment from the other overload

smowton reviewed (commented):
Ahhh, I get it, I hadn't noticed the single-arg version of `increase_generation` is only being used in one place (`symex_dead.cpp`). In that case rename that overload "`increase_generation_if_exists` and then this LGTM

smowton commented:
Renamed the PR to reflect the situation a little more accurately

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 871909f71109ff6cf2c2522fd2bb6f5d09debb5d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102358182

tautschnig commented:
@JohnDumbell This is failing CI on Windows.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 73363ab8e31a233a52d387585d8f0bb20e675551).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102377249

tautschnig commented on src/goto-symex/path_storage.h:
```diff
@@ -90,6 +90,13 @@ class path_storaget
+  /// Tracks fresh L2 names across an entire symex run. This will be shared
+  /// across all states to allocate unique generations.
+  std::shared_ptr<global_name_mapt> level2_names =
+    std::make_shared<global_name_mapt>(global_name_mapt());
```
Why is this a shared pointer rather than a plain member?

tautschnig commented on src/goto-symex/renaming_level.cpp:
```diff
@@ -77,6 +77,45 @@ void symex_level1t::restore_from(
+  current_names.emplace(l1_identifier, std::make_pair(lhs, 0));
+  global_names->emplace(l1_identifier, std::make_pair(lhs, 0));
+
+  increase_generation_if_exists(l1_identifier);
```
I think there needs to be third, private, function that both `increase_generation` and `increase_generation_if_exists` invoke. That function should take iterators into `current_names` and `global_names`. As is, we first do an `emplace` here and then again do lookups inside `increase_generation_if_exists` (as well as a whole lot of redundant invariant checking).

JohnDumbell replied:
In the end I just duplicated the common code in each method as it was just two lines.

tautschnig commented on src/goto-symex/path_storage.h:
```diff
@@ -90,6 +90,13 @@ class path_storaget
   /// Counter for nondet objects, which require unique names
   symex_nondet_generatort build_symex_nondet;
 
+  using global_name_mapt = std::map<irep_idt, std::pair<ssa_exprt, unsigned>>;
```
Just a comment about the commit message: please wrap lines at 80 or 72 characters, otherwise `git log` becomes unusable.

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -221,7 +221,16 @@ void goto_symext::symex_goto(statet &state)
     state.top().goto_state_map[new_state_pc];
 
-  goto_state_list.emplace_back(state);
+  // If the guard is true (and the alternative branch unreachable) we can move
```
But ... we pass `state` to `symex_transition` just below?!

smowton replied:
This suggests a constructor taking an ordinary ref and a boolean `can_move_state` or something, which partially destroys `goto_symex_statet` (but critically leaves its source and stats unchanged / suitable for continued usage)?

JohnDumbell replied:
I've added this change in as an additional commit, please check it when able as I wasn't sure what fields were safe to move or not.

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -514,10 +523,16 @@ static void merge_names(
+    dest_state.level2 = std::move(goto_state.level2);
+    dest_state.propagation = std::move(goto_state.propagation);
+    return;
+  }
```
I'm not sure `phi_function` is the right place for this, I'd do it at the caller (and not invoke `phi_function`).

tautschnig reviewed (commented):
What is the performance impact of this change?

smowton commented:
Benchmark for some Tika methods (mostly sameish, one rather slower, could be noise): 
![perf_out](https://user-images.githubusercontent.com/283483/53506985-fe7f8000-3aae-11e9-9372-b3782c9333e0.png)

smowton commented:
Will try to construct an example showing off the specific benefit now...

smowton commented:
* Needs clang-format
* Comment still outstanding at `symex_goto.cpp:224`
* Move `increase_generation`, `increase_generation_if_exists` from `symex_level2t` to `goto_symex_statet`

tautschnig commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -318,6 +320,8 @@ class goto_symex_statet final : public goto_statet
   bool run_validation_checks;
 
 private:
+  path_storaget &path_storage;
```
Is this actually being used? Certainly not in this commit as far as I can tell?!

JohnDumbell replied:
It only got actively used in the next commit, but I've squashed them both together now as it makes more sense.

tautschnig commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -439,10 +437,7 @@ bool goto_symex_statet::l2_thread_read_encoding(
-        level2.current_names.emplace(l1_identifier, std::make_pair(ssa_l1, 0))
-          .first;
-      symex_renaming_levelt::increase_counter(level2_it);
+      level2.increase_generation(l1_identifier, ssa_l1, path_storage);
```
I think you should pass through `path_storage` as an argument, not make it a member of `goto_symex_statet`. This will increase the churn I'm afraid, but otherwise the coupling becomes too difficult to manage. While doing so, you actually shouldn't pass in all of `path_storage`, but just the bit that you really need.

smowton replied:
@tautschnig you'll then need to pass it to every call, and every recursive call, to `rename`, since that calls thread-read-encoding which in turn allocates a fresh L2 name. Are you sure that's better?

tautschnig replied:
That's what we are doing with field sensitivity... I actually do think that's better, but if you don't want to go as far then at least only a very limited part of `path_storaget` should be passed in. Otherwise we're creating a monolith. It's the sort of thing that @LAJW and @romainbrenguier have been arguing against in #4282.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 74aaf2b9df46fb68ab84e66010f4cab71cdfb6ba).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102692673
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 5541ce592cec5a5dd8ae7c21b3fad14332c43ffe).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102705449
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

JohnDumbell commented:
I've applied comments so far and fixed up CI. Joel-bot failures don't seem related to what I've been doing, so I'll just wait on those.

tautschnig commented on src/goto-symex/renaming_level.h:
```diff
@@ -18,6 +18,8 @@ Author: Romain Brenguier, romain.brenguier@diffblue.com
 #include <util/irep.h>
 #include <util/ssa_expr.h>
 
+class path_storaget;
```
Is this actually used?

tautschnig commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -322,6 +326,16 @@ class goto_symex_statet final : public goto_statet
+  /// isn't an expression keyed by it.
+  void increase_generation_if_exists(const irep_idt identifier);
+
+  global_name_mapt &global_name_map;
```
Is this a public member?

tautschnig commented on src/goto-symex/path_storage.h:
```diff
@@ -90,6 +90,12 @@ class path_storaget
   /// Counter for nondet objects, which require unique names
   symex_nondet_generatort build_symex_nondet;
 
+  using global_name_mapt = std::map<irep_idt, std::pair<ssa_exprt, unsigned>>;
```
`path_storage.h` already includes `goto_symex_state.h`, thus there is no need to repeated the declaration.

tautschnig commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -220,7 +220,17 @@ void goto_symext::symex_goto(statet &state)
+  else
+  {
+    goto_state_list.emplace_back(state);
+  }
```
We need to find a better way for doing this: we now have, with only a `symex_transition` in between, the same `if(new_guard.is_true())` branching structure. This is hard to reason about for compilers and humans alike.

tautschnig reviewed (commented):
This is looking a lot better I think, but I'm still waiting for performance data that shows the benefit of this change. The proposed changes do complicate the overall architecture because of the shared state (the numbering map), so we should really only be doing this if we have performance data backing it.

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -316,16 +326,26 @@ void goto_symext::merge_gotos(statet &state)
+    phi_function(goto_state, state);
+  }
 
   merge_value_sets(goto_state, state);
```
Apologies for late comment, just occurred to me now -- `merge_value_sets` should be moving when we have `dest.guard.is_false()` as well.

tautschnig replied:
Well, if only we had #4169 ...

smowton reviewed (commented):
Noticed missing move-from `goto_statet::value_set`. To this end, `merge_goto(...)` should take a `goto_statet &&`, since that thing really is dying (`merge_gotos` is just about to delete the original from `goto_state_map`)

JohnDumbell committed:
Merge 'if new_guard == true' blocks

This means that the symex_transition call is duplicated, but better
that one call than multiple if statements.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 96230eb92687ad481ae0970e83c816abd7430f1c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102839061

smowton commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -512,6 +510,43 @@ bool goto_symex_statet::l2_thread_read_encoding(
+  // If we have a global store, increment its generation count, then assign
+  // that new value to our local scope.
+  auto global_names_iter = global_name_map.find(identifier);
+  if(global_names_iter != global_name_map.end())
```
Found locally but not globally should be impossible -- use `.at`.

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -210,17 +210,24 @@ void goto_symext::symex_goto(statet &state)
-  symex_transition(state, state_pc, backward);
-
-  // adjust guards
+  // If the guard is unconditionally true it means the alternative branch is
```
Not so -- (1) there is no one unique alternative branch, there can be several paths around this code; (2) this isn't the state guard, it's the GOTO condition. `new_guard.is_true()` just means an unconditional jump, regardless of whether there are paths around this code.

Suggested phrasing: // On an unconditional GOTO we don't need our state any more, as it will be overwritten by `merge_goto`. Therefore we move it onto `goto_state_list` instead of copying as usual.

smowton reviewed (approved):
Two more TODOs:
1. Clang-format / lint
2. Double-check if any of the added `= default` constructors are no longer necessary now that we have the partial-move constructor and `level2_renamingt` is not being changed.

tautschnig commented:
I do find the "partial move" thingy very subtle, but can live with it. So I'm not opposed to the changes, but I am still wondering what the performance gain is? Or at least I don't see any reason for making these changes if it isn't to improve performance.

JohnDumbell commented:
I'll let you know when I have some concrete numbers - some basic tests last week showed about a 10% improvement to runtime but I want to do more extreme cases and see what happens there.

tautschnig commented:
Given that #3980 has been merged, is there any chance to actually re-use the `path_storaget::get_unique_index` that it introduced?

JohnDumbell commented:
Sure, I'll have a look and see how it can fit in

smowton committed:
Symex: protect level2

We really should protect the level 2 name map to prevent "allocating" a new generation
by just making one up in the local name map, cf. applying for one properly using
increase_generation.

smowton commented:
@tautschnig candidate branch that styles this more like your L1-naming patch here: https://github.com/diffblue/cbmc/compare/develop...smowton:jd/enhancement/symex_state_moving

Note I added one more commit to protect `level2`, on grounds that making up a new name by just directly writing to it is now dangerous (you must use the proper allocator). You might want to do something similar to the L1 map for similar reasons.

smowton commented:
@tautschnig behold: data exhibiting the branch above doing its thing!

Test program:
```java
public class Test {

    public static void main(int unknown) {

	// Step 1: allocate thousands of objects, so the level2
	// name map / constant propagator are quite large
	for(int i = 0; i < 1000; i++) {
	    Object o = new Object();
	}

	// Step 2: Make a few branches in a context which isn't
	// certain to run (i.e. where the state guard is non-trivial)
	if(unknown % 5 == 0) {
	    int x = 0;
	    for(int j = 0; j < 1000; j++) {
		if(x == 0)
		    x = 0;
		else
		    x = 1;
	    }
	}
       
    }

}
```
With develop: 5.3s
With this change: 0.8s

The reason: the `if(x == 0)` in the step 2 loop is decided statically, and at the end of the `if` block we unconditionally jump to the end of the loop body. Because the state guard isn't `true` we suppose there might be a route around this code (there is, but it's outside the loop). Therefore we (move, with this PR | copy, without this PR) the state to the end-of-loop instruction (the unconditional back-edge, in fact) and walk over each intermediate instruction checking there are no incoming edges. Similarly when we get to that back-edge we (copy | move) the state from the `goto_state_map`.

Conclusion: the 2000x copies of an L2 name map containing around 1000 objects takes around 4s longer than 2000x moves. I expect this to provide a handsome speedup when symexing code which is largely constant (so branches are statically decided) in a context where there are thousands of values in the L2 name map / constant propagator.

tautschnig commented:
@JohnDumbell @smowton Nice results! Now we just need to get this PR in a shape where CI passes :-)

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b09cd6493ef54f6308465577c48df87b53ad805e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104080211

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 440eb97cfe8b8e727dbaa17b4bf8401f12be1b7e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/104131701

tautschnig commented:
Unless you have decent access to a Windows development environment, my (rather simplistic) strategy to debug the segmentation fault on Windows would be to push reverts of subsets of the commits to identify which of the commits is causing the problem.

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -210,17 +210,24 @@ void goto_symext::symex_goto(statet &state)
+  // instead of copying as usual.
   if(new_guard.is_true())
   {
+    goto_state_list.emplace_back(state.source, std::move(state));
```
Definitely comment here that that `std::move` is moving our `goto_statet` superclass, rather than the whole of `state`

smowton commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -319,14 +326,21 @@ void goto_symext::merge_goto(
+  {
+    if(state.guard.is_false())
+    {
+      state.move_from(std::move(goto_state));
```
I think this is now replacable by the default move-assign op? (Again comment that this is our superclass being moved)

JohnDumbell replied:
Kinda, see https://github.com/diffblue/cbmc/pull/4199/commits/f5ae411ea6c81ee96457906a08cf1e065dd565ac for one way to do it.

smowton replied:
That looks good to me -- it'd be cleaner if it was a member not a superclass, but otherwise as hoped for.

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -500,6 +493,32 @@ bool goto_symex_statet::l2_thread_read_encoding(
+  const ssa_exprt &lhs)
+{
+  auto current_emplace_res =
+    level2.current_names.emplace(l1_identifier, std::make_pair(lhs, 0));
```
It looks like this method belongs to symex_level2t. ￼￼:leaves:

smowton replied:
This was considered but not done because we didn't want `symex_level2t` concerned with the global name allocator (only with the path-local map)

romainbrenguier commented on src/goto-symex/goto_symex_state.cpp:
```diff
@@ -500,6 +493,32 @@ bool goto_symex_statet::l2_thread_read_encoding(
+{
+  // If we can't find the name in the local scope, this is a no-op.
+  auto current_names_iter = level2.current_names.find(identifier);
+  if(current_names_iter == level2.current_names.end())
```
￼:leaves:

romainbrenguier commented on src/goto-symex/goto_symex_state.h:
```diff
@@ -202,7 +205,17 @@ class goto_symex_statet final : public goto_statet
+  void increase_generation_if_exists(const irep_idt identifier);
+
 private:
+  std::function<std::size_t(const irep_idt &)> fresh_l2_name_provider;
```
Could maybe be part of `symex_level2t`. I'm a bit worried when I see `goto_symex_statet` grow, as this class already contains too many thing and does too much to my taste.

romainbrenguier commented on src/goto-symex/path_storage.h:
```diff
@@ -115,8 +116,24 @@ class path_storaget
+    name_index_mapt &unique_index_map,
+    const irep_idt &id,
+    std::size_t minimum_index)
+  {
```
definition should go to the cpp file. Also I would suggest declaring a class for `name_index_mapt` (instead of a typedef) and make this the only public method.

smowton replied:
Seems short and hot enough to be worth inlining to me

romainbrenguier commented on src/goto-symex/symex_goto.cpp:
```diff
@@ -210,7 +210,17 @@ void goto_symext::symex_goto(statet &state)
-  goto_state_list.emplace_back(state.source, state);
+
+  // If the guard is true (and the alternative branch unreachable) we can move
+  // the state in this case as it'll never be accessed on the alternate branch.
```
"in this case" seems redundant in this sentence

romainbrenguier commented on src/goto-symex/goto_state.h:
```diff
@@ -28,8 +28,15 @@ class goto_statet
   unsigned depth = 0;
 
+protected:
   symex_level2t level2;
```
What about making current_names private in symex levels instead? (or both)

smowton replied:
Both sounds reasonable to me, but suggest doing further refactoring elsewhere, this PR has already dragged on a fair while
